@shubham-pampattiwar unit test
```
/go/src/github.com/openshift/oadp-operator/bin/oadp-dev/golangci-lint run
internal/controller/velero_test.go:3522:1: File is not properly formatted (gci)
						"app.kubernetes.io/managed-by": "user-manager", // should be ignored
^
1 issues:
```

--- a/internal/controller/nodeagent.go
+++ b/internal/controller/nodeagent.go
@@ -9,6 +9,10 @@ import (
 
 	"github.com/go-logr/logr"
 	configv1 "github.com/openshift/api/config/v1"
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	"github.com/openshift/oadp-operator/pkg/common"
+	"github.com/openshift/oadp-operator/pkg/credentials"
+	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
 	"github.com/operator-framework/operator-lib/proxy"
 	"github.com/vmware-tanzu/velero/pkg/install"
 	"github.com/vmware-tanzu/velero/pkg/util/kube"
@@ -21,11 +25,6 @@ import (
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-
-	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
-	"github.com/openshift/oadp-operator/pkg/common"
-	"github.com/openshift/oadp-operator/pkg/credentials"
-	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
 )
 
 const (
@@ -509,6 +508,17 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
 	// Note: NOT applied to Spec.Selector.MatchLabels as those are immutable after creation
 	ds.Labels = applyResourceLabels(dpa, ds.Labels)
 	ds.Spec.Template.Labels = applyResourceLabels(dpa, ds.Spec.Template.Labels)
+
+	// Re-assert selector labels to ensure template labels match selector
+	// This prevents user resourceLabels from overriding selector-critical labels
+	if ds.Spec.Selector != nil && ds.Spec.Selector.MatchLabels != nil {
+		if ds.Spec.Template.Labels == nil {
+			ds.Spec.Template.Labels = make(map[string]string)
+		}
+		for k, v := range ds.Spec.Selector.MatchLabels {
+			ds.Spec.Template.Labels[k] = v
+		}
+	}
 
 	// Apply user-provided resource annotations to both daemonset and pod template
 	ds.Annotations = applyResourceAnnotations(dpa, ds.Annotations)

--- a/internal/controller/nonadmin_controller.go
+++ b/internal/controller/nonadmin_controller.go
@@ -8,6 +8,8 @@ import (
 	"time"
 
 	"github.com/go-logr/logr"
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	"github.com/openshift/oadp-operator/pkg/common"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/exp/maps"
 	appsv1 "k8s.io/api/apps/v1"
@@ -18,9 +20,6 @@ import (
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-
-	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
-	"github.com/openshift/oadp-operator/pkg/common"
 )
 
 const (
@@ -139,6 +138,17 @@ func (r *DataProtectionApplicationReconciler) buildNonAdminDeployment(deployment
 	// Note: NOT applied to Spec.Selector.MatchLabels as those are immutable after creation
 	deploymentObject.Labels = applyResourceLabels(r.dpa, deploymentObject.Labels)
 	deploymentObject.Spec.Template.Labels = applyResourceLabels(r.dpa, deploymentObject.Spec.Template.Labels)
+
+	// Re-assert selector labels to ensure template labels match selector
+	// This prevents user resourceLabels from overriding selector-critical labels
+	if deploymentObject.Spec.Selector != nil && deploymentObject.Spec.Selector.MatchLabels != nil {
+		if deploymentObject.Spec.Template.Labels == nil {
+			deploymentObject.Spec.Template.Labels = make(map[string]string)
+		}
+		for k, v := range deploymentObject.Spec.Selector.MatchLabels {
+			deploymentObject.Spec.Template.Labels[k] = v
+		}
+	}
 
 	// Apply user-provided resource annotations to both deployment and pod template
 	deploymentObject.Annotations = applyResourceAnnotations(r.dpa, deploymentObject.Annotations)

--- a/internal/controller/velero.go
+++ b/internal/controller/velero.go
@@ -9,6 +9,11 @@ import (
 
 	"github.com/go-logr/logr"
 	"github.com/google/go-cmp/cmp"
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	"github.com/openshift/oadp-operator/pkg/common"
+	"github.com/openshift/oadp-operator/pkg/credentials"
+	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
+	veleroserver "github.com/openshift/oadp-operator/pkg/velero/server"
 	"github.com/operator-framework/operator-lib/proxy"
 	"github.com/sirupsen/logrus"
 	"github.com/vmware-tanzu/velero/pkg/install"
@@ -25,12 +30,6 @@ import (
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-
-	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
-	"github.com/openshift/oadp-operator/pkg/common"
-	"github.com/openshift/oadp-operator/pkg/credentials"
-	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
-	veleroserver "github.com/openshift/oadp-operator/pkg/velero/server"
 )
 
 const (
@@ -855,10 +854,13 @@ func filterOutResourceLabels(dpa *oadpv1alpha1.DataProtectionApplication, labels
 		result[k] = v
 	}
 
-	// Remove resourceLabels if present
+	// Remove resourceLabels if present, but skip protected labels
+	// Protected labels should always be preserved in matchLabels
 	if dpa != nil && dpa.Spec.ResourceLabels != nil {
 		for k := range dpa.Spec.ResourceLabels {
-			delete(result, k)
+			if !isProtectedLabel(k) {
+				delete(result, k)
+			}
 		}
 	}
 

--- a/internal/controller/velero.go
+++ b/internal/controller/velero.go
@@ -188,6 +188,10 @@ func (r *DataProtectionApplicationReconciler) buildVeleroDeployment(veleroDeploy
 		uploaderType = dpa.Spec.Configuration.NodeAgent.UploaderType
 	}
 
+	// Filter out resourceLabels before passing to install.Deployment
+	// This ensures matchLabels don't contain resourceLabels (which would cause immutable selector errors on reconcile)
+	labelsForInstall := filterOutResourceLabels(dpa, veleroDeployment.Labels)
+
 	installDeployment := install.Deployment(veleroDeployment.Namespace,
 		install.WithResources(veleroResourceReqs),
 		install.WithImage(getVeleroImage(dpa)),
@@ -195,7 +199,7 @@ func (r *DataProtectionApplicationReconciler) buildVeleroDeployment(veleroDeploy
 		install.WithFeatures(dpa.Spec.Configuration.Velero.FeatureFlags),
 		install.WithUploaderType(uploaderType),
 		// last label overrides previous ones
-		install.WithLabels(veleroDeployment.Labels),
+		install.WithLabels(labelsForInstall),
 		// use WithSecret false even if we have secret because we use a different VolumeMounts and EnvVars
 		// see: https://github.com/vmware-tanzu/velero/blob/ed5809b7fc22f3661eeef10bdcb63f0d74472b76/pkg/install/deployment.go#L223-L261
 		// our secrets are appended to containers/volumeMounts in credentials.AppendPluginSpecificSpecs function
@@ -234,7 +238,10 @@ func (r *DataProtectionApplicationReconciler) customizeVeleroDeployment(veleroDe
 	if veleroDeployment.Spec.Selector.MatchLabels == nil {
 		veleroDeployment.Spec.Selector.MatchLabels = make(map[string]string)
 	}
-	veleroDeployment.Spec.Selector.MatchLabels, err = common.AppendUniqueKeyTOfTMaps(veleroDeployment.Spec.Selector.MatchLabels, veleroDeployment.Labels, getDpaAppLabels(dpa))
+	// Filter out resourceLabels from veleroDeployment.Labels before adding to matchLabels
+	// matchLabels are immutable, so we must not include resourceLabels which are user-configurable
+	labelsForMatchLabels := filterOutResourceLabels(dpa, veleroDeployment.Labels)
+	veleroDeployment.Spec.Selector.MatchLabels, err = common.AppendUniqueKeyTOfTMaps(veleroDeployment.Spec.Selector.MatchLabels, labelsForMatchLabels, getDpaAppLabels(dpa))
 	if err != nil {
 		return fmt.Errorf("velero deployment selector label: %v", err)
 	}
@@ -830,6 +837,28 @@ func applyResourceLabels(dpa *oadpv1alpha1.DataProtectionApplication, coreLabels
 	for k, v := range dpa.Spec.ResourceLabels {
 		if !isProtectedLabel(k) {
 			result[k] = v
+		}
+	}
+
+	return result
+}
+
+// filterOutResourceLabels removes resourceLabels from the given labels map.
+// This is used to get labels that are safe for matchLabels (which are immutable).
+func filterOutResourceLabels(dpa *oadpv1alpha1.DataProtectionApplication, labels map[string]string) map[string]string {
+	if labels == nil {
+		return nil
+	}
+
+	result := make(map[string]string)
+	for k, v := range labels {
+		result[k] = v
+	}
+
+	// Remove resourceLabels if present
+	if dpa != nil && dpa.Spec.ResourceLabels != nil {
+		for k := range dpa.Spec.ResourceLabels {
+			delete(result, k)
 		}
 	}
 

--- a/internal/controller/velero_test.go
+++ b/internal/controller/velero_test.go
@@ -21,6 +21,10 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/onsi/ginkgo/v2"
 	"github.com/onsi/gomega"
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	oadpclient "github.com/openshift/oadp-operator/pkg/client"
+	"github.com/openshift/oadp-operator/pkg/common"
+	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
 	"github.com/operator-framework/operator-lib/proxy"
 	"github.com/sirupsen/logrus"
 	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
@@ -33,11 +37,6 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
-	oadpclient "github.com/openshift/oadp-operator/pkg/client"
-	"github.com/openshift/oadp-operator/pkg/common"
-	"github.com/openshift/oadp-operator/pkg/credentials/stsflow"
 )
 
 const (
@@ -3519,8 +3518,8 @@ func TestApplyResourceLabels(t *testing.T) {
 				Spec: oadpv1alpha1.DataProtectionApplicationSpec{
 					ResourceLabels: map[string]string{
 						"app.kubernetes.io/name":       "user-override", // should be ignored
-						"app.kubernetes.io/managed-by": "user-manager", // should be ignored
-						"custom-label":                 "custom-value", // should be included
+						"app.kubernetes.io/managed-by": "user-manager",  // should be ignored
+						"custom-label":                 "custom-value",  // should be included
 					},
 				},
 			},
@@ -3529,9 +3528,9 @@ func TestApplyResourceLabels(t *testing.T) {
 				"app.kubernetes.io/managed-by": common.OADPOperator,
 			},
 			expectedLabels: map[string]string{
-				"app.kubernetes.io/name":       "velero",       // core label preserved
+				"app.kubernetes.io/name":       "velero",            // core label preserved
 				"app.kubernetes.io/managed-by": common.OADPOperator, // core label preserved
-				"custom-label":                 "custom-value", // user label added
+				"custom-label":                 "custom-value",      // user label added
 			},
 		},
 		{

--- a/internal/controller/vmfilerestore_controller.go
+++ b/internal/controller/vmfilerestore_controller.go
@@ -7,6 +7,8 @@ import (
 	"strconv"
 
 	"github.com/go-logr/logr"
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	"github.com/openshift/oadp-operator/pkg/common"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/exp/maps"
 	appsv1 "k8s.io/api/apps/v1"
@@ -19,9 +21,6 @@ import (
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-
-	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
-	"github.com/openshift/oadp-operator/pkg/common"
 )
 
 const (
@@ -140,6 +139,17 @@ func (r *DataProtectionApplicationReconciler) buildVMFileRestoreDeployment(deplo
 	// Note: NOT applied to Spec.Selector.MatchLabels as those are immutable after creation
 	deploymentObject.Labels = applyResourceLabels(r.dpa, deploymentObject.Labels)
 	deploymentObject.Spec.Template.Labels = applyResourceLabels(r.dpa, deploymentObject.Spec.Template.Labels)
+
+	// Re-assert selector labels to ensure template labels match selector
+	// This prevents user resourceLabels from overriding selector-critical labels
+	if deploymentObject.Spec.Selector != nil && deploymentObject.Spec.Selector.MatchLabels != nil {
+		if deploymentObject.Spec.Template.Labels == nil {
+			deploymentObject.Spec.Template.Labels = make(map[string]string)
+		}
+		for k, v := range deploymentObject.Spec.Selector.MatchLabels {
+			deploymentObject.Spec.Template.Labels[k] = v
+		}
+	}
 
 	// Apply user-provided resource annotations to both deployment and pod template
 	deploymentObject.Annotations = applyResourceAnnotations(r.dpa, deploymentObject.Annotations)

